include/oneapi/dpl/pstl/hetero/dpcpp/sycl_traits.h

```diff
@@ -0,0 +1,152 @@
+_ONEDPL_DEVICE_COPYABLE(__transform_if_unary_functor)
+_ONEDPL_DEVICE_COPYABLE(__transform_if_binary_functor)
+_ONEDPL_DEVICE_COPYABLE(__replace_functor)
+_ONEDPL_DEVICE_COPYABLE(__replace_copy_functor)
+_ONEDPL_DEVICE_COPYABLE(zip_forward_iterator)
+
+template <typename _SourceT>
+struct fill_functor;
+
+template <typename _Generator>
+struct generate_functor;
+
+template <typename _Pred>
+struct equal_predicate;
+
+template <typename _Tp, typename _Pred>
+struct __search_n_unary_predicate;
+
+template <typename _Predicate>
+struct adjacent_find_fn;
+
+template <class _Comp>
+struct __is_heap_check;
+
+template <typename _Predicate, typename _ValueType>
+struct __create_mask_unique_copy;
+
+_ONEDPL_DEVICE_COPYABLE(fill_functor)
+_ONEDPL_DEVICE_COPYABLE(generate_functor)
+_ONEDPL_DEVICE_COPYABLE(__brick_fill)
+_ONEDPL_DEVICE_COPYABLE(__brick_fill_n)
+_ONEDPL_DEVICE_COPYABLE(__search_n_unary_predicate)
+_ONEDPL_DEVICE_COPYABLE(__is_heap_check)
+_ONEDPL_DEVICE_COPYABLE(equal_predicate)
+_ONEDPL_DEVICE_COPYABLE(adjacent_find_fn)
+_ONEDPL_DEVICE_COPYABLE(__create_mask_unique_copy)
+
+using namespace oneapi::dpl::__par_backend_hetero;
+
+template <typename _ExecutionPolicy, typename _Pred>
+struct __early_exit_find_or;
+
+_ONEDPL_DEVICE_COPYABLE(__early_exit_find_or);
+
+using namespace oneapi::dpl::unseq_backend;
+
+_ONEDPL_DEVICE_COPYABLE(walk_n)
+_ONEDPL_DEVICE_COPYABLE(walk_adjacent_difference)
+_ONEDPL_DEVICE_COPYABLE(transform_reduce)
+_ONEDPL_DEVICE_COPYABLE(reduce_over_group)
+_ONEDPL_DEVICE_COPYABLE(single_match_pred_by_idx)
+_ONEDPL_DEVICE_COPYABLE(multiple_match_pred)
+_ONEDPL_DEVICE_COPYABLE(n_elem_match_pred)
+_ONEDPL_DEVICE_COPYABLE(first_match_pred)
+_ONEDPL_DEVICE_COPYABLE(__create_mask)
+_ONEDPL_DEVICE_COPYABLE(__copy_by_mask)
+_ONEDPL_DEVICE_COPYABLE(__partition_by_mask)
+_ONEDPL_DEVICE_COPYABLE(__global_scan_functor)
+_ONEDPL_DEVICE_COPYABLE(__scan)
+_ONEDPL_DEVICE_COPYABLE(__brick_includes)
+_ONEDPL_DEVICE_COPYABLE(__brick_set_op)
+_ONEDPL_DEVICE_COPYABLE(__brick_reduce_idx)
+
+using namespace oneapi::dpl::internal;
+
+template <typename Comp, typename T, int func>
+struct custom_brick;
+
+template <typename T, typename Predicate>
+struct replace_if_fun;
+
+template <typename T, typename Predicate, typename UnaryOperation>
+class transform_if_stencil_fun;
+
+template <typename ValueType, typename FlagType, typename BinaryOp>
+struct segmented_scan_fun;
+
+template <typename Output1, typename Output2>
+class scatter_and_accumulate_fun;
+
+template <typename ValueType, typename FlagType, typename BinaryOp>
+struct scan_by_key_fun;
+
+_ONEDPL_DEVICE_COPYABLE(custom_brick)
+_ONEDPL_DEVICE_COPYABLE(replace_if_fun)
+_ONEDPL_DEVICE_COPYABLE(scan_by_key_fun)
+_ONEDPL_DEVICE_COPYABLE(segmented_scan_fun)
+_ONEDPL_DEVICE_COPYABLE(scatter_and_accumulate_fun)
+_ONEDPL_DEVICE_COPYABLE(transform_if_stencil_fun)
+
+using namespace oneapi::dpl;
+
+_ONEDPL_DEVICE_COPYABLE(zip_iterator)
+_ONEDPL_DEVICE_COPYABLE(transform_iterator)
```

Comment:
For `transform_iterator`, and `permutation_iterator`, it is possible we need to adjust our strategy slightly.  I think we may want to wait for our upcoming meeting before merging this PR.
Otherwise, I believe this LGTM.